### ISSUE TITLE
Determine parallelization strategy for pathway Enrichr

### ISSUE THREAD

**Issue by matthewhirschey** [matthewhirschey/ddh.org#16]
Need to determine best computational strategy to get pathways. Discussed with @johnbradley to split gene list into deciles, for 10 possible parallel queries

**Comment by matthewhirschey** [comment]
My email to Avi:

> Avi,
> I’ve enjoyed using the Enrichr resource you have developed, as well as the R package to make it accessible programmatically — as a side note, thank you for your help on Twitter when an update to R broke EnrichR. 
> 
> I’ve been working hard to develop a new resource that measures patterns of gene similarity across several data sources. I was hoping to take the gene lists generated fro my analysis and run them through EnrichR to measure a wide-range of gene pathway associations; but want to be cognizant of your server load. 
> 
> Would it be OK to run 10 queries in parallel to run through my gene lists? I didn’t see any information (i.e. robot.txt) about requests for limiting number of queries or rate, so thought I’d ask. 
> We can also make sure to do this in off hours. 
> 
> Thanks for making great resources available to the research community,
> Matt

**Comment by matthewhirschey** [comment]
Avi's response:

> Hi Matthew,
> 
> It would be great if you can place some delay (0.5 sec) between requests if possible so you do not overwhelm the server.
> 
> Best regards,
> 
> Avi

**Comment by matthewhirschey** [comment]
I’m not sure I understand.
Can we run 10 parallel jobs as long as there is a delay between each request to enrichr made by a single job?
Or do we need to run one long job that has a delay between each request it makes?
 
@johnbradley

**Comment by matthewhirschey** [comment]
My read is that we can run 10 in parallel, but need to put a delay between each request.

@johnbradley: Do you want me to explore furrr? Or do you have another idea?

**Comment by johnbradley** [comment]
One concern I have with having a single R script that runs in parallel is recovering from an error.
If one thread/process fails an hour into running you may have to start the whole thing over if you don’t have robust retry logic.
If it’s broken up into separate processes that run independently we can just rerun those that failed.
My plan was to change generate_depmap_pathways.R to write out per-decile files, add another R script to merge them and change the Makefile to wrap this up.

If you would rather try out using furrr that’s fine with me. It may be that furrr is a more typical R approach to this problem.

**Comment by matthewhirschey** [comment]
I think your approach sounds better. I had thought about the failure, and unsure about how to recover, etc. 

NB. `furrr` is a new package so certainly not "typical"

@johnbradley Go a head and give it a shot

**Comment by johnbradley** [comment]
I was able to run the pathway Enrichr code in parallel creating smaller files then merging them into master_negative.RData and master_positive.RData. The results are uploaded to the DDS `ddh-data` project. These files seem to work with the shiny app. I did have to modify `19Q3_expression_join.RData` to copy the 'stripped_cell_line_name' field to `cell_line`.

I hit a bug when generating bottom pathway enrichment related to genes that have no correlated genes once after filtering `achilles_lower`. I have a fix in my branch, but I will create a separate issue to determine the best way to address it.

I hope to get my branch into a PR the beginning of this week.

**Comment by matthewhirschey** [comment]
re: "These files seem to work with the shiny app. I did have to modify 19Q3_expression_join.RData to copy the 'stripped_cell_line_name' field to cell_line."

I updated the generate_data.R (in cd49a6ee, ready to push) to change the name to `cell_line`
